# Método de Riemann

Este método consiste en tomar un intervalo [a , b] del dominio de una función f y generar particiones de este intervalo, es decir, sea f: [a , b] -> R  una función acotada, entonces podemos aproximar el área bajo la curva  generada por esta función por medio de rectángulos, sea xk un elemento del intervalo [a , b], donde a < xk < b, y sea t1 elemento del intervalo [a , xk] y t2 elemento de [xk , b], podemos hacer una primera aproximación con dos rectángulos uno de base (xk – a) y altura f(t1) y un segundo rectángulo de base (b – xk) y altura f(t2), esto nos daría una primera aproximación al área bajo la curva de f, donde:

                    A = f(t1)(xk – a) + f(t2)(b – xk)
                
Es natural pensar que entre mayor número de particiones sean mejor será la aproximación, por lo tanto podemos construir un número finito de particiones.

Sea {x0, x1 , ….. , xn}, donde a = x0 < x1 < x2 < …..< xk < ……< xn = b y sea tk elemento de [xk-1 , xk] para toda k= 1, …., n. Entonces podemos generar un rectángulo para cada sub intervalo de base (xk  - x(k-1)) y altura f(tk), entonces tenemos que:
               A = f(t1)(x1-a) +...+ f(tk)(xk  - x(k-1)) +…+f(tn)(b-x(n-1)) = ∑ f(tk)(xk  - x(k-1)), desde k = 1 hasta k = n

En el límite cuando n -> ∞ A = ∫f(x)dx integrando desde a hasta b

In [7]:
function riemann(f::Function,a,b,N)
    x=linspace(a,b,N)
    g=0
    for i in 2:N
        dx=x[i]-x[i-1]
        h=f(x[i-1])
        g=(h*dx)+g
    end
    return g
end


riemann (generic function with 1 method)

In [19]:
f(x)= cos(x)

f (generic function with 1 method)

In [54]:
riemann(f,0,pi/2,10000)

1.0000785456145262

# Regla del trapecio 

La regla del trapecio es muy similar al método de Riemann la diferencia es que en lugar de utilizar rectángulos se utilizan trapecios, el método consiste en:

Se traza una recta entre los puntos (a , f(a)) y (b, f(b)) y se calcula el área de este trapecio de base menor (b-a) y base mayor (f(a) – f(b)), al igual que en el método anterior entre mayor sea el número de particiones mejor será la aproximación si partimos el int. [a, b] en n sub intervalos de ancho h = (b – a)/2 podemos ver que:

La integral de a hasta b de f(x) es: ∫〖f(x)dx 〗≈h/2[f(a)+2f(a+h)+2f(a+2h)+⋯+f(b)] 

Con un error aproximado Er = - ((b-a)3/12n2)f’’(α), con α un elemento de [a, b].

In [46]:
function trapecios(f::Function,a,b,N)
    x=linspace(a,b,N)
    h=0
    for i in 2:N
        dx=x[i]-x[i-1]
        g=(f(x[i-1])+f(x[i]))/2
        h=(g*dx)+h
    end
    return h
end


trapecios (generic function with 1 method)

In [62]:
trapecios(f,0,pi/2,10000)

0.999999997943423

# Regla de Simpson de orden 2

Este método interpola una función con un polinomio de Lagrange de orden 2, Sea f : [a, b] -> R y sea x1 un elemento de [a , b], entonces podemos ver a I = ∫f(x)dx  de a hasta b como la siguiente integral:

        I= ∫〖[(x-x1)(x-b)/(a-x1)(a-b) 〗 f(a)+(x-a)(x-b)/(x1-a)(x1-b)  f(x1)+(x-x1)(x-a)/(b-a)(b-x1)  f(x1)]dx

Integrando desde a hasta b:

            I≅h/6[f(a)+4f(x1)+f(b)] , con h =(b – a) y x1 = (a+b)/2

In [60]:
function simpson(f::Function,a,b)
    h=(b-a)/6
    x1 = (a+b)/2
    I=(h)*(f(a)+f(b)+(4*f(x1)))
    return I
    
end

simpson (generic function with 3 methods)

In [61]:
simpson(f,0,pi/2)

1.0022798774922104

# Test

Para probar las funciones se utilizó la funcion coseno, sabemos que la integral del coseno desde 0 a pi/2 es igual a 1, y como podemos ver en los resultados son números mui cercanos al cero.

riemann(cos(x),0,pi/2,10000) = 1.0000785456145262

trapecios(cos(x),0,pi/2,10000) = 0.999999997943423

simpson(f,0,pi/2) = 1.0022798774922104